readthedocs/projects/tasks/utils.py

```diff
@@ -321,6 +323,49 @@ def deprecated_config_file_used_notification():
     )
 
 
+@app.task(queue="web")
+def set_builder_scale_in_protection(instance, protected_from_scale_in):
+    """
+    Set scale-in protection on this builder ``instance``.
+
+    This way, AWS will not scale-in this instance while it's building the documentation.
+    This is pretty useful for long running tasks.
+    """
+    log.bind(instance=instance, protected_from_scale_in=protected_from_scale_in)
+
+    if settings.DEBUG or settings.RTD_DOCKER_COMPOSE:
+        log.info(
+            "Running development environment. Skipping scale-in protection.",
+        )
+        return
+
+    asg = boto3.client(
+        "autoscaling",
+        aws_access_key_id=settings.RTD_AWS_SCALE_IN_ACCESS_KEY,
+        aws_secret_access_key=settings.RTD_AWS_SCALE_IN_SECRET_ACCESS_KEY,
+        region_name=settings.RTD_AWS_SCALE_IN_REGION_NAME,
+    )
+
+    # web-extra-i-0c3e866c4e323928f
+    hostname_match = re.match(r"([a-z\-]+)-(i-[a-z0-9]+)", instance)
```

Comment:
Not likely that any match would be greedier than we want here, but EC2 identifiers are hex encoded:

```suggestion
    hostname_match = re.match(r"([a-z\-]+)-(i-[a-f0-9]+)", instance)
```